Add `InitReqType` to classify `FrameRefType` by init/reset requirements.

bjoeris opened the pull request:
## Description

This change introduces a new enum `InitReqType` that describes how a resource needs to be initialized/reset for replay, according to its `FrameRefType`.

Commit:
Add `InitReqType` to classify `FrameRefType` by init/reset requirements.

The accumulated `FrameRefType` of the accesses to a resource within a frame
determines how the resource needs to be initialized. The possibilities are:

* `None`: The initial value of the resource is never used within the frame,
  either because it is never read, or because it is only read after being
  cleared
* `InitOnce`: The resource needs to be initialized before replaying the frame
  for the first time, but does not ever need to be reset; this occurs when a
  resource is read within the frame, but not written within the frame.
* `Reset`: The resource needs to be reset each time before replaying the frame.
  This occurs when the resource is read and later written within the frame.

baldurk commented on renderdoc/core/resource_manager.h:
```diff
@@ -114,6 +114,39 @@ FrameRefType ComposeFrameRefsUnordered(FrameRefType first, FrameRefType second);
+{
+  // Initial contents of the resource are not used, and need not be initialized.
+  // Corresponds to `None`, `Write` or `Clear` ref types.
+  eInitReq_None,
```
With the `Clear` ref type once that is reached then it's a sink state so we never change.

That means if I follow correctly that you could have operations that are (coarsely): `Clear`, `Read`, `Write`, `Read`. In which case we _do_ need to reset so that the third `Write` isn't visible to the earlier second `Read`. It's just that the reset is cheap because it's just a clear rather than copying data. You could also have `Clear` then only `Read`s, which means we only need to clear just once at startup.

It feels like clear is parallel to whether we need to init once or reset each time. In my head there are five cases to consider: `InitOnceByClearing`, `ResetByClearing`, `InitOnceFromData`, `ResetFromData`, in ascending expense. Then finally a `None` for resources that are truly untouched or are only written to without being read from (and so in theory can be left alone).

Does that track with your mental model? It seems like this `None` is overloaded a bit by containing `Clear`.

bjoeris replied:
Hmm, I think something is not matching up in the way we are thinking about `Clear`. In my model, a resource being in the `Clear` state means that the first use *in the frame* was a `Clear`. So when we replay the frame, the initial content of the resource will be completely overwritten by the `Clear` command before any other commands try to read the resource.

In your `Clear`, `Read`, `Write`, `Read` example, the third `Write` (in replay `N`) will not be visible to the second `Read` (in replay `N+1`), because the `Clear` (in replay `N+1`) will be executed prior to that read.

I did not intend the name "Clear" to indicate the manner in which the resource is init or reset, only to convey that the initial content has been entirely overwritten.

baldurk replied:
Ahh I see what you mean, yes. In your example if you don't clear at the initial state then you'd still get leaking data if the user examined the resource at a point in the frame before the clear happened though, immediately after having replayed later in the frame after some writes, is that correct? That's something we'd want to avoid I think otherwise you run the risk of misleading the user with false data. At least if the buffer/texture is completely nulled out it's clearer that there is no previous data.

I think there's a bit of aliasing in terminology coming up here. I wonder if there's some way to disambiguate - maybe rename `Clear` to `CompleteWrite` and `Write` to `PartialWrite`? I'm not sure if that's any improvement but at least it gets us away from clear which is something we might still want to do (as above).

bjoeris replied:
I had not considered the possible confusion of examining the resource before the `Clear`, and seeing data written later in the frame. Perhaps we could store when the resource was first used, which would let us detect attempts to read the undefined contents. Is there a similar problem for resources created within the frame, being examined before they are created?

I think that optimizing out resets of `Clear`ed resources was a significant part of the improved performance in code gen optimizations, so I hope we can find a way to do the same in native replay.

`CompleteWrite` and `PartialWrite` do sound better in this case. Do you think this terminology would be confusing if we do a "Complete" write to a sub-interval of a memory, or of some subresource of an image?

baldurk replied:
I think currently there's no problem for resources created within a frame, because one of the few uses of the frame reference types currently is that any resource which is ever marked for `Write` will have initial states *created* for it with `Create_InitialState`, then applied in the same fashion as if they had been saved. This is intended to cover precisely this hole where something is written within the frame so needs a reset, but didn't have initial states saved. Typically for resources created mid-frame it's just a fast clear, especially on vulkan.

With your changes I'd wonder how much of the improved performance was down to removing redundant _copies_ of `Clear`ed resources specifically, and how much was down to removing those resets generally? From what you'd described before I thought most of the benefit came from the first case - I'd expect a GPU to be very fast at clearing an image or buffer region while I could understand accumulated cost from doing redundant copies of (worst case) GBs of data that hasn't changed.

We could in theory optimise by detecting the first write and selectively clearing only when we hit that region, but I'd want to know that it's actually a significant performance saving first. My gut feeling is that clearing images/buffer regions without any copy should be fast enough - a few milliseconds even is a generous figure and shouldn't be significant in UI terms.

(Side-note: I think we'll want to eventually have a default-off option to disable these optimisations that technically change behaviour, and keep initial contents even for `Clear`ed resources. While the conservative approach you're taking means that missing data before a `Clear` would never actually impact the frame itself, I've found that people often want to be able to see these non-semantic parts of a capture. Ironically this would probably just be restoring the 'save all initials' option that was previously D3D11-only and removed in v1.2).

----

I want to say that `CompleteWrite` and `PartialWrite` shouldn't be confusing as long as you have the context that it's applying to a sub-interval. No matter what terminology we pick, if one of the states is intended to mean "this is completely overwritten before ever being read, and nothing of the previous contents is visible afterwards" then you have to have context that it only applies to a specific subsection and not to the whole resource. To me the core distinction is that - whether or not any previous writes are visible - so complete/partial were the terms that came to mind but there might be a better alternative.

I know nitpicking over naming/terminology can be tedious, so apologies!

Commit:
Rename `FrameRefType` values to clarify meanings

Commit:
Added `Clear` value to `InitReqType`

This value is similar to `None`, in that the resource need not be initialized
for correct replay; however, a resource with `Clear` type might be modified in
the frame, where a resource with `None` type may not be modified in the frame.

This distinction may be helpful in initializing the irrelevant initial contents
to some known value (e.g. 0), so that users inspecting these resources don't see
unexpected values (such as a value written later in the frame). `None` resources
need only set this dummy initial contents once, where `Clear` resources need to
reset this dummy initial contents every frame.

baldurk commented:
Looks good now, at least to the point where further bikeshedding won't help :). Thanks for the clarification. I'll merge when CI is green.
